tests/vm_userdata_test.go

```diff
@@ -56,71 +70,71 @@ var _ = Describe("CloudInit UserData", func() {
 		_, ok := obj.(*v1.VM)
 		Expect(ok).To(BeTrue(), "Object is not of type *v1.VM")
 		tests.WaitForSuccessfulVMStart(obj)
-	}
 
-	Context("CloudInit Data Source NoCloud", func() {
-		It("should launch multiple VMs with cloud-init data source NoCloud", func(done Done) {
-			num := 2
-			vms := make([]*v1.VM, 0, num)
-			objs := make([]runtime.Object, 0, num)
-			for i := 0; i < num; i++ {
-				vm, err := tests.NewRandomVMWithUserData("kubevirt/cirros-registry-disk-demo:devel", "noCloud")
+		ws := Dial(vm.ObjectMeta.GetName(), "serial0")
+		defer ws.Close()
+		next := ""
+		Eventually(func() string {
+			for {
+				if index := strings.Index(next, "\n"); index != -1 {
+					line := next[0:index]
+					next = next[index+1:]
+					return line
+				}
+				_, data, err := ws.ReadMessage()
 				Expect(err).ToNot(HaveOccurred())
-				obj := LaunchVM(vm)
-				vms = append(vms, vm)
-				objs = append(objs, obj)
+				next = next + string(data)
 			}
+		}, 45*time.Second).Should(ContainSubstring(magicStr))
+	}
 
-			for idx, vm := range vms {
-				VerifyUserDataVM(vm, objs[idx])
-			}
+	BeforeEach(func() {
+		tests.BeforeTestCleanup()
+	})
 
+	Context("CloudInit Data Source NoCloud", func() {
+		It("should launch vm with cloud-init data source NoCloud", func(done Done) {
+			vm, err := tests.NewRandomVMWithUserData("noCloud", userData)
+			Expect(err).ToNot(HaveOccurred())
+			obj := LaunchVM(vm)
+			VerifyUserDataVM(vm, obj)
```

Comment:
Could you pass in the magic string to this function? It's quite hard to follow against what you are verifying otherwise.

Reply:
@rmohr userData and magicStr and now something that are generated in the same place the vm is created, launched, and verified from.   I agree this makes the tests a little easier to follow.